src/engine/QueryExecutionTree.cpp

```diff
@@ -399,38 +355,66 @@ nlohmann::json QueryExecutionTree::writeQLeverJsonTable(
   return json;
 }
 
+// _____________________________________________________________________________
+
+vector<std::optional<pair<size_t, ResultTable::ResultType>>>
+QueryExecutionTree::buildValidIndices(const vector<string>& selectVars,
+                                      const ResultTable& resultTable) const {
+  vector<std::optional<pair<size_t, ResultTable::ResultType>>> validIndices;
+  for (auto var : selectVars) {
+    constexpr std::string_view prefix = "TEXT(";
+    constexpr size_t prefixLength = prefix.length();
+    if (var.starts_with(prefix)) {
+      var = var.substr(prefixLength, var.rfind(')') - prefixLength);
+    }
+    auto it = getVariableColumns().find(var);
+    if (it != getVariableColumns().end()) {
+      validIndices.emplace_back(pair<size_t, ResultTable::ResultType>(
+          it->second, resultTable.getResultType(it->second)));
+    } else {
+      validIndices.emplace_back(std::nullopt);
+    }
+  }
+  return validIndices;
+}
+
 // _____________________________________________________________________________
 template <QueryExecutionTree::ExportSubFormat format>
 ad_utility::stream_generator::stream_generator QueryExecutionTree::writeTable(
-    size_t from, size_t upperBound,
-    const vector<std::optional<pair<size_t, ResultTable::ResultType>>>
-        validIndices,
-    std::shared_ptr<const ResultTable> resultTable) const {
+    const vector<string>& selectVars, size_t limit, size_t offset) const {
   static_assert(format == ExportSubFormat::BINARY ||
                 format == ExportSubFormat::CSV ||
                 format == ExportSubFormat::TSV);
-  if (!resultTable) {
-    resultTable = getResult();
-  }
+
+  // They may trigger computation (but does not have to).
+  shared_ptr<const ResultTable> resultTable = getResult();
+  LOG(DEBUG) << "Resolving strings for finished binary result...\n";
+  auto validIndices = buildValidIndices(selectVars, *resultTable);
+
   const auto& idTable = resultTable->_idTable;
+  size_t upperBound = std::min<size_t>(offset + limit, idTable.size());
+
   // special case : binary export of IdTable
   if constexpr (format == ExportSubFormat::BINARY) {
-    for (size_t i = from; i < upperBound; ++i) {
-      for (size_t j = 0; j < validIndices.size(); ++j) {
-        if (validIndices[j]) {
-          const auto& val = *validIndices[j];
+    for (size_t i = offset; i < upperBound; ++i) {
```

Comment:
Now that we have `libstdc++` working again, we could systematically use `std::views::iota`
(basically python `range()` if we want to.